wikitools/code_block_parser.py

```diff
@@ -0,0 +1,117 @@
+import typing
+
+
+class CodeTag(typing.NamedTuple):
+    start: int
+    len: int
+
+
+class CodeBlock(typing.NamedTuple):
+    """
+    A Markdown code block (monospace font + gray backround). Could be one of the following:
+        - Inline: `text` or `` text `` (the latter is used for snippets containing `)
+        - Multiline: ``` followed by anything else until the next ``` block
+
+    Examples:
+
+        ```test test test```
+        ^ start            ^ end
+
+        `this is a very long phrase`
+        ^ start                    ^ end
+
+    Similarly to the HTML comments, links in code blocks are not accounted for.
+    """
+
+    start: int  # 0-based position of the first character of the block's opening tag
+    end: int  # 0-based position of the last character of the block's closing tag
+    tag_len: int
+
+    @property
+    def is_multiline(self):
+        return self.end == -1 and self.tag_len == 3
+
+    def contains(self, other: 'CodeBlock'):
+        return (
+            self.start < other.start and
+            self.end > other.end and
+            self.tag_len > other.tag_len
+        )
+
+
+class CodeBlockParser:
+    def __init__(self):
+        self.__in_multiline = False
+
+    @property
+    def in_multiline(self) -> bool:
+        return self.__in_multiline
+
+    def parse(self, line: str) -> typing.List[CodeBlock]:
+        blocks: typing.List[CodeBlock] = []
+
+        tag_stack = []
+        if self.__in_multiline:
+            tag_stack.append(CodeTag(start=-1, len=3))
+
+        i = 0
+        while i < len(line):
+            if line[i] != '`':
+                i += 1
+                continue
+
+            cnt = 0
+            while i + cnt < len(line) and line[i + cnt] == '`':
+                cnt += 1
+
+            # just closed the code block
+            if tag_stack and tag_stack[-1].len == cnt:
+                opening_tag = tag_stack.pop()
+                # add the new code block
+                # if it's mistakenly included into a smaller one (` ```test``` `), remark will catch that anyway
```

Comment:
smaller tags can contain bigger tags like this `` ` ```test``` ` `` --> ` ```test``` `

in fact any tag pair can contain any other tags as long as they're of different lengths

well at least here on github, not sure what remark says about it

Reply:
yep that's my understanding. removed the comment and fixed that part